### PR TITLE
Avoid unnecessary temporaries when ref'ing a DST value

### PR DESCRIPTION
**Pull request by dotdash**
A DST value and a fat pointer to it have the same representation, all we
have to do is to adjust the type of the datum holding the pointer.

### CODE REVIEW

File: src/librustc_trans/trans/expr.rs [hunk 1]
```diff
@@ -1709,8 +1699,11 @@ fn trans_addr_of<'blk, 'tcx>(bcx: Block<'blk, 'tcx>,
     let mut bcx = bcx;
     let sub_datum = unpack_datum!(bcx, trans_to_lvalue(bcx, subexpr, "addr_of"));
     if !type_is_sized(bcx.tcx(), sub_datum.ty) {
-        // DST lvalue, close to a fat pointer
-        ref_fat_ptr(bcx, sub_datum)
+        let dest_ty = bcx.tcx().mk_imm_ref(bcx.tcx().mk_region(ty::ReStatic), sub_datum.ty);
```
Comment (arielb1) [inline src/librustc_trans/trans/expr.rs]: `expr_ty`?

Comment (dotdash) [inline src/librustc_trans/trans/expr.rs]: Thanks, fixed!

**Comment by brson** [comment]
Codegen improvements. Any numbers?

**Comment by dotdash** [comment]
Pretty negligible as far as optimized builds are concerned AFAICT. Didn't
bother to check unoptimized builds as this seems to be rather uncommon in
general. Just had me confused when looking at specific IR results.
Am 02.10.2015 20:27 schrieb "Brian Anderson" <notifications@github.com>:

**Pull request merged**